File: prolly_tree.go
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"strconv"
	"fmt"
)

// GreaterThan checks if one node is 'greater than' another.
func (n *Node) GreaterThan(other *Node) bool {
    return n.isTail || n.timestamp > other.timestamp
}

// LessThan checks if one node is 'less than' another.
func (n *Node) LessThan(other *Node) bool {
    return (!n.isTail) && n.timestamp < other.timestamp
}

// EqualTo checks if one node is 'equal to' another.
func (n *Node) EqualTo(other *Node) bool {
    return n.timestamp == other.timestamp
}

// NotEqualTo checks if one node is 'not equal to' another.
func (n *Node) NotEqualTo(other *Node) bool {
    return n.timestamp != other.timestamp
}

// GreaterThanOrEqualTo checks if one node is 'greater than or equal to' another.
func (n *Node) GreaterThanOrEqualTo(other *Node) bool {
    return n.isTail || n.timestamp >= other.timestamp
}

// LessThanOrEqualTo checks if one node is 'less than or equal to' another.
func (n *Node) LessThanOrEqualTo(other *Node) bool {
    return (!n.isTail) && n.timestamp <= other.timestamp
}

// CalculateHash computes the SHA256 hash of a given string.
func CalculateHash(hashableStr string) string {
	hasher := sha256.New()
	hasher.Write([]byte(hashableStr))
	return hex.EncodeToString(hasher.Sum(nil))
}

// IsBoundaryHash checks if the last hex digit of the hash is below a certain threshold.
func IsBoundaryHash(hashValue string, threshold int) bool {
	// Convert the last hex digit of the hash to an integer.
	lastHexDigit := hashValue[len(hashValue)-1:]
	hashInt, _ := strconv.ParseInt(lastHexDigit, 16, 64)
	return int(hashInt) < threshold
}

// BucketHash calculates the Merkel hash of a bucket of nodes by concatenating their Merkel hashes.
func BucketHash(nodes []*Node) string {
	// Concatenate the merkel hashes of the nodes.
	var hashStr string
	for _, node := range nodes {
		hashStr += node.merkelHash
	}
	// Calculate the hash of the concatenated string.
	return CalculateHash(hashStr)
}

// Node represents each node in the tree structure.
type Node struct {
	timestamp   int
	data        string
	nodeHash    string
	level       int
	up          *Node
	down        *Node
	left        *Node
	right       *Node
	merkelHash  string
	boundary    *bool // using a pointer to differentiate between unset and false
	isTail      bool
}

// NewNode creates and initializes a new Node.
func NewNode(data string, timestamp int, isTail bool) *Node {
	n := &Node{
		timestamp:  timestamp,
		data:       data,
		isTail:     isTail,
		level:      0,
		up:         nil,
		down:       nil,
		left:       nil,
		right:      nil,
		merkelHash: CalculateHash(data + strconv.Itoa(timestamp)),
		nodeHash:   CalculateHash(data + strconv.Itoa(timestamp)),
		boundary:   nil,
	}
	return n
}

// CreateHigherLevelNode creates a new node one level higher.
func (n *Node) CreateHigherLevelNode() *Node {
	newNode := NewNode("", n.timestamp, n.isTail)
	newNode.level = n.level + 1
	newNode.down = n
	n.up = newNode
	newNode.nodeHash = CalculateHash(n.nodeHash)
	return newNode
}

// FillMerkelHash fills the merkel hash of the node by going down walking left till you hit the boundary.
func (n *Node) FillMerkelHash() {
	var bucketNodes []*Node
	node := n.down
	bucketNodes = append(bucketNodes, node)

	for node.left != nil {
		if node.left.IsBoundaryNode() {
			break
		}
		node = node.left
		bucketNodes = append(bucketNodes, node)
	}

	for _, node := range bucketNodes {
		if node.merkelHash == "" {
			node.FillMerkelHash()
		}
	}

	// Reverse the slice of nodes.
	for i, j := 0, len(bucketNodes)-1; i < j; i, j = i+1, j-1 {
		bucketNodes[i], bucketNodes[j] = bucketNodes[j], bucketNodes[i]
	}

	n.merkelHash = BucketHash(bucketNodes)
}

// IsBoundaryNode checks if the node is a boundary node.
func (n *Node) IsBoundaryNode() bool {
	if n.boundary != nil {
		return *n.boundary
	}
	boundary := n.isTail || IsBoundaryHash(n.nodeHash, 7) // Assuming the threshold is 7
	n.boundary = &boundary
	return boundary
}

// FindNextBoundaryNode finds the next boundary node by going right till you hit the boundary node.
func (n *Node) FindNextBoundaryNode() *Node {
	node := n
	for node.right != nil {
		if node.right.IsBoundaryNode() {
			return node.right
		}
		node = node.right
	}
	return node
}

// Level represents a level in the ProllyTree.
type Level struct {
	level int
	tail  *Node
}

// NewLevel creates a new Level instance.
func NewLevel(lvl int) *Level {
	return &Level{level: lvl}
}

// ToList converts a level's nodes to a list.
func (l *Level) ToList() []*Node {
	var nodes []*Node
	node := l.tail
	for node != nil {
		nodes = append([]*Node{node}, nodes...) // Prepend to reverse
		node = node.left
	}
	return nodes
}

// String provides a string representation of the level.
func (l *Level) String() string {
	var timestamps []int
	for _, node := range l.ToList() {
		timestamps = append(timestamps, node.timestamp)
	}
	return " <=> " + fmt.Sprint(timestamps)
}

// BaseLevel creates a base level from a list of messages.
func BaseLevel(messages []*Message) *Level {
	level := NewLevel(0)
	var nodes []*Node
	for _, m := range messages {
		nodes = append(nodes, NewNode(m.data, m.timestamp, false))
	}
	fakeTail := NewNode("Tail", -1, true)
	nodes = append(nodes, fakeTail)
	level.tail = LinkNodes(nodes)[len(nodes)-1]
	return level
}

// NextLevel creates a level from the previous level.
func NextLevel(prevLevel *Level) *Level {
	nodes := prevLevel.ToList()
	var eligibleNodes []*Node
	for _, node := range nodes {
		if node.IsBoundaryNode() {
			eligibleNodes = append(eligibleNodes, node.CreateHigherLevelNode())
		}
	}
	linkedNodes := LinkNodes(eligibleNodes)
	for _, node := range linkedNodes {
		node.FillMerkelHash()
	}
	newLevel := NewLevel(prevLevel.level + 1)
	newLevel.tail = linkedNodes[len(linkedNodes)-1]
	return newLevel
}

// LinkNodes links nodes together in a level.
func LinkNodes(nodes []*Node) []*Node {
	for i := 0; i < len(nodes)-1; i++ {
		nodes[i].right = nodes[i+1]
		nodes[i+1].left = nodes[i]
	}
	return nodes
}

// ProllyTree represents the entire tree structure.
type ProllyTree struct {
	levels []*Level
}

// NewProllyTree creates a new ProllyTree instance with the given messages.
func NewProllyTree(messages []*Message) *ProllyTree {
	tree := &ProllyTree{}
	baseLevel := BaseLevel(messages)
	tree.levels = append(tree.levels, baseLevel)
	for len(baseLevel.ToList()) > 1 {
		baseLevel = NextLevel(baseLevel)
		tree.levels = append(tree.levels, baseLevel)
	}
	return tree
}

// Insert adds a message to the ProllyTree.
func (pt *ProllyTree) Insert(message *Message) {
	newNode := NewNode(message.data, message.timestamp, false)
	rightOfNewNode := pt.findNodeGreaterThan(newNode.timestamp)
	pt.insertNodeAtLevel(newNode, 0, rightOfNewNode)
}

// Delete removes a node from the ProllyTree based on the timestamp.
func (pt *ProllyTree) Delete(timestamp int) *Node {
	originalNode := pt.Search(timestamp)
	node := originalNode
	if originalNode == nil {
		return nil
	}
	rightBoundaryNode := originalNode.FindNextBoundaryNode()

	for node != nil {
		leftNode := node.left
		rightNode := node.right
		if rightNode != nil {
			rightNode.left = leftNode
		}
		if leftNode != nil {
			leftNode.right = rightNode
		}
		node = node.up
	}

	// Clean up empty levels
	var levelsToRemove []int
	for i := len(pt.levels) - 1; i > 1; i-- {
		levelBelow := pt.levels[i-1]
		if levelBelow.tail.left == nil { // Only has tail node
			levelsToRemove = append(levelsToRemove, i)
		}
	}

	for _, levelIndex := range levelsToRemove {
		pt.levels = append(pt.levels[:levelIndex], pt.levels[levelIndex+1:]...)
	}

	pt.updatePropagateMerkelHash(rightBoundaryNode.up)

	return originalNode
}

// Search finds a node based on the timestamp.
func (pt *ProllyTree) Search(timestamp int) *Node {
	rightNode := pt.findNodeGreaterThan(timestamp)
	if rightNode.left != nil && rightNode.left.timestamp == timestamp {
		return rightNode.left
	}
	return nil
}

// String provides a string representation of the ProllyTree.
func (pt *ProllyTree) String() string {
	var levels []string
	for _, level := range pt.levels {
		levels = append(levels, level.String())
	}
	return fmt.Sprint(levels)
}

// findNodeGreaterThan finds the node with the smallest timestamp that is greater than the given timestamp.
func (pt *ProllyTree) findNodeGreaterThan(timestamp int) *Node {
	nodeToFind := NewNode("", timestamp, false)
	node := pt.levels[len(pt.levels)-1].tail // Start from the root

	// Move down to the level 0 right boundary node of the given timestamp
	for node.down != nil {
		if node.left != nil && node.left.timestamp >nodeToFind.timestamp {
			node = node.left
		} else {
			node = node.down
		}
	}

	// Move left until the left element is the timestamp or lower than the timestamp
	for node.left != nil && node.left.timestamp >nodeToFind.timestamp {
		node = node.left
	}

	return node
}

// insertNodeAtLevel inserts a node at a given level and creates a new level if necessary.
func (pt *ProllyTree) insertNodeAtLevel(newNode *Node, levelIndex int, rightOfNewNode *Node) {
	if levelIndex >= len(pt.levels) {
		panic("Adding on a level that doesn't exist")
	}

	// Insert the new node to the right of the rightOfNewNode
	ourLeftNode := rightOfNewNode.left
	rightOfNewNode.left = newNode
	newNode.right = rightOfNewNode
	newNode.left = ourLeftNode
	if ourLeftNode != nil {
		ourLeftNode.right = newNode
	}

	if newNode.IsBoundaryNode() {
		higherLevelNode := newNode.CreateHigherLevelNode()
		nextBoundaryNode := newNode.FindNextBoundaryNode()
		if levelIndex == len(pt.levels)-1 {
			pt.addEmptyLevel()
		}
		pt.insertNodeAtLevel(higherLevelNode, levelIndex+1, nextBoundaryNode.up)
	}

	lastLevel := pt.levels[len(pt.levels)-1]

	if len(lastLevel.ToList()) > 1 {
		pt.addEmptyLevel()
	}
}

// updatePropagateMerkelHash updates the Merkel hash of the node and propagates it up.
func (pt *ProllyTree) updatePropagateMerkelHash(node *Node) {
	node.FillMerkelHash()
	if node.up == nil {
		if node.isTail {
			return
		}
		nextBoundaryNode := node.FindNextBoundaryNode()
		pt.updatePropagateMerkelHash(nextBoundaryNode.up)
	} else {
		pt.updatePropagateMerkelHash(node.up)
	}
}

func (pt *ProllyTree) addEmptyLevel() {
	levelIndex := len(pt.levels)
	newLevel := NewLevel(levelIndex)
	pt.levels = append(pt.levels, newLevel)
	upgradedTail := pt.levels[levelIndex-1].tail.CreateHigherLevelNode()
	newLevel.tail = upgradedTail
}

// Message represents the data structure to be added to the tree.
type Message struct {
	data      string
	timestamp int
}

// NewMessage creates and initializes a new Message.
func NewMessage(data string, timestamp int) *Message {
	return &Message{data: data, timestamp: timestamp}
}

// String provides a string representation of the message, similar to Python's __repr__.
func (m *Message) String() string {
	return fmt.Sprintf("Message(%s, %s)", m.data, m.timestamp)
}

// to Test the prolly tree
func main() {
    fmt.Println("Creating 9 messages: ")
	var messages []*Message
	for i := 0; i < 10; i++ {
		messages = append(messages, NewMessage(strconv.Itoa(i), i))
	}

	fmt.Println("Creating a prolly tree with the messages")
	tree := NewProllyTree(messages)

	fmt.Println("Printing the tree")
	fmt.Println(tree)

	fmt.Println("#############################################")
	fmt.Println("Searching a node with timestamp 5")
	foundNode := tree.Search(5)
	if foundNode != nil {
		fmt.Println("Key Found inside the tree: Node(", foundNode.data, ",", foundNode.timestamp, ")")
	} else {
		fmt.Println("Key not found in the tree.")
	}

	fmt.Println("#############################################")
	fmt.Println("Deleting a node with timestamp 6")
	tree.Delete(6)
	fmt.Println("Printing the tree")
	fmt.Println(tree)

	fmt.Println("#############################################")
	fmt.Println("Inserting a node with timestamp 12")
	tree.Insert(NewMessage("12", 12))
	fmt.Println("Printing the tree")
	fmt.Println(tree)
}

File: prolly_tree.nim
import tables, strutils, math, std/strutils, algorithm
import
  sequtils,
  nimcrypto/sha2,
  nimcrypto/utils

type
  Node* = ref object
    data, nodeHash, merkelHash: string
    timestamp, level: int
    up, down, left, right: Node
    boundary: bool
    boundarySet: bool
    isTail: bool

  Message* = object
    data: string
    timestamp: int

  Level* = ref object
    level: int
    tail: Node

  ProllyTree* = ref object
    levels: seq[Level]

proc greaterThan*(n: Node, other: Node): bool =
  ## Checks if one node is 'greater than' another.
  return n.isTail or n.timestamp > other.timestamp

# Define the methods on the Node type
proc newMessage*(data: string, timestamp: int): Message =
  result.data = data
  result.timestamp = timestamp

proc calculateHash*(hashableStr: string): string =
  var ctx: sha256
  ctx.init()  # Initialize the context
  ctx.update(hashableStr)  # Update the context with the input string
  let digest = ctx.finish()  # Finalize the context and get the digest

  # Convert the digest to a hexadecimal string using the nimcrypto utils
  result = toHex(digest.data).toLowerAscii()

proc bucketHash(nodes: seq[Node]): string =
  ## Calculate the Merkel hash of a bucket of nodes.
  ## The bucket is a sequence of nodes. The Merkel hash is calculated
  ## by concatenating the Merkel hashes of the nodes in the bucket
  ## and then calculating the hash of the concatenated string.
  
  var concatenatedHashes = ""
  for node in nodes:
    concatenatedHashes &= node.merkelHash  # Concatenate the Merkel hashes.

  return calculateHash(concatenatedHashes)

proc isBoundaryHash(hashValue: string, threshold = 7): bool =
  # Check if the last hex digit of the hash is below the threshold
  let lastChar = hashValue[hashValue.len - 1]
  # convert the last hex digit to an integer
  var hashInt:int = fromHex[int]($lastChar)
  result = hashInt < threshold

proc newNode*(data: string, timestamp: int, isTail: bool): Node =
  new(result)
  result.timestamp = timestamp
  result.data = data
  result.isTail = isTail
  result.nodeHash = calculateHash(data & $timestamp)
  result.merkelHash = calculateHash(data & $timestamp)
  result.up = nil
  result.down = nil
  result.left = nil
  result.right = nil
  result.boundary = false  # Initialize to false.
  result.boundarySet = false  # Initialize to false, indicating boundary not explicitly set.

proc createHigherLevelNode*(n: Node): Node =
  let newNode = newNode("", n.timestamp, n.isTail)
  newNode.level = n.level + 1
  newNode.down = n
  n.up = newNode
  newNode.nodeHash = calculateHash($(n.nodeHash))
  return newNode

proc isBoundaryNode*(n: Node): bool =
  if not n.boundarySet:  # Check if the boundary has been explicitly set.
    # Calculate the boundary value based on the node's properties.
    n.boundary = n.isTail or isBoundaryHash(n.nodeHash)
    n.boundarySet = true  # Mark the boundary as explicitly set.
  return n.boundary

proc findNextBoundaryNode*(n: Node): Node =
  var node = n
  while node.right != nil:
    if node.right.isBoundaryNode:
      return node.right
    node = node.right
  return node

proc newLevel*(lvl: int): Level =
  new(result)
  result.level = lvl

proc toList*(l: Level): seq[Node] =
  result = @[]
  var node = l.tail
  
  while node != nil:
    result.add(node)
    node = node.left

  # Reverse the sequence in place
  reverse(result)

proc `$`*(l: Level): string =
  result = " <=> " & $l.toList().mapIt(it.timestamp)

proc linkNodes*(nodes: seq[Node]): seq[Node] =
  for i in 0..<nodes.len-1:
    nodes[i].right = nodes[i+1]
    nodes[i+1].left = nodes[i]
  return nodes

proc baseLevel*(messages: seq[Message]): Level =
  let level = newLevel(0)
  var nodes: seq[Node]
  for m in messages:
    nodes.add(newNode(m.data, m.timestamp, false))
  let fakeTail = newNode("Tail", -1, true)
  nodes.add(fakeTail)
  level.tail = linkNodes(nodes)[^1]
  return level

proc fillMerkelHash*(n: Node) =
  if n.isNil:
    return

  var bucketNodes: seq[Node]  # Make sure the sequence can hold mutable Nodes  
  var node = n.down

  if node != nil:
    bucketNodes.add(node)  # Ensure that 'node' is mutable when added
  
  while node != nil and node.left != nil:
    if node.left.isBoundaryNode():
      break
    node = node.left  # Ensure 'left' returns a mutable Node if it's supposed to be modified
    bucketNodes.add(node)  # Ensure that 'node' is mutable when added

  for node in bucketNodes:
    if node.merkelHash == "":
      fillMerkelHash(node)  # This should be fine if 'node' is indeed a mutable Node

  # Reverse the sequence of nodes.
  reverse(bucketNodes)

  n.merkelHash = bucketHash(bucketNodes)  # Ensure that 'bucketHash' can accept the modified sequence

proc nextLevel*(prevLevel: Level): Level =
  let nodes = prevLevel.toList()
  var eligibleNodes: seq[Node]
  for node in nodes:
    if node.isBoundaryNode:
      eligibleNodes.add(node.createHigherLevelNode())
  let linkedNodes = linkNodes(eligibleNodes)
  for node in linkedNodes:
    node.fillMerkelHash()
  let newLevel = newLevel(prevLevel.level + 1)
  newLevel.tail = linkedNodes[^1]
  return newLevel


proc newProllyTree*(messages: seq[Message]): ProllyTree =
  ## Creates a new ProllyTree instance with the given messages.
  new(result)  # Allocate memory for the ProllyTree.
  result.levels = @[]  # Initialize levels as an empty sequence.
  
  var level = baseLevel(messages)  # Create the base level.
  result.levels.add(level)  # Add the base level to the tree.
  
  while level.toList().len > 1:  # While the level has more than one node.
    level = nextLevel(level)  # Create the next level.
    result.levels.add(level)
    

proc findNodeGreaterThan*(pt: ProllyTree, timestamp: int): Node =
  let nodeToFind = newNode("", timestamp, false)
  var node = pt.levels[^1].tail  # Start from the root

  # Move down to the level 0 right boundary node of the given timestamp
  while not node.down.isNil:
    if not node.left.isNil and node.left.greaterThan(nodeToFind):
      node = node.left
    else:
      node = node.down

  # Move left until the left element is the timestamp or lower than the timestamp
  while not node.left.isNil and node.left.greaterThan(nodeToFind):
    node = node.left
  #  print a deubg statement and print the node timestamp
  return node

proc addEmptyLevel*(pt: ProllyTree) =
  ## Adds an empty level to the ProllyTree.
  let levelIndex = pt.levels.len
  let newLevel = newLevel(levelIndex)  # Create a new level.
  pt.levels.add(newLevel)  # Add the new level to the tree.

  if pt.levels[levelIndex - 1].tail != nil:  # Check if the previous level's tail exists.
    let upgradedTail = createHigherLevelNode(pt.levels[levelIndex - 1].tail)  # Create a higher-level node from the tail.
    newLevel.tail = upgradedTail  # Set the new level's tail to the upgraded tail.

proc insertNodeAtLevel*(pt: ProllyTree, newNode: Node, levelIndex: int, rightOfNewNode: Node) =
  ## Inserts a node at a given level and creates a new level if necessary.
  if levelIndex >= len(pt.levels):
    raise newException(ValueError, "Adding on a level that doesn't exist")

  # Insert the new node to the right of the rightOfNewNode
  let ourLeftNode = rightOfNewNode.left
  rightOfNewNode.left = newNode
  newNode.right = rightOfNewNode
  newNode.left = ourLeftNode
  if ourLeftNode != nil:
    ourLeftNode.right = newNode

  if newNode.isBoundaryNode():  # Assuming 'isBoundaryNode' is a method or property of Node.
    let higherLevelNode = newNode.createHigherLevelNode()  # Assuming 'createHigherLevelNode' is a method of Node.
    let nextBoundaryNode = newNode.findNextBoundaryNode()  # Assuming 'findNextBoundaryNode' is a method of Node.
    if levelIndex == len(pt.levels) - 1:
      pt.addEmptyLevel()  # Assuming 'addEmptyLevel' is a method of ProllyTree.
    pt.insertNodeAtLevel(higherLevelNode, levelIndex + 1, nextBoundaryNode.up)

  let lastLevel = pt.levels[len(pt.levels) - 1]

  if len(lastLevel.toList()) > 1:  # Assuming 'toList' is a method or property of Level.
    pt.addEmptyLevel()

proc updatePropagateMerkelHash*(pt: ProllyTree, node: Node) =
  if node.isNil:
    return
  node.fillMerkelHash()
  
  if node.up.isNil:
    if node.isTail:
      return
    let nextBoundaryNode = node.findNextBoundaryNode()
    pt.updatePropagateMerkelHash(nextBoundaryNode.up)
  else:
    pt.updatePropagateMerkelHash(node.up)

proc insert*(pt: ProllyTree, message: Message) =
  let newNode = newNode(message.data, message.timestamp, false)
  let rightOfNewNode = pt.findNodeGreaterThan(newNode.timestamp)
  pt.insertNodeAtLevel(newNode, 0, rightOfNewNode)

proc search*(pt: ProllyTree, timestamp: int): Node =
  let rightNode = pt.findNodeGreaterThan(timestamp)
  if not rightNode.left.isNil and rightNode.left.timestamp == timestamp:
    return rightNode.left
  return nil

proc delete*(pt: ProllyTree, timestamp: int): Node =
  
  let originalNode = pt.search(timestamp)
  var node = originalNode
  if originalNode.isNil:
    return nil

  let rightBoundaryNode = originalNode.findNextBoundaryNode()
  
  while not node.isNil:
    var leftNode = node.left
    var rightNode = node.right
    if not rightNode.isNil:
      rightNode.left = leftNode
    if not leftNode.isNil:
      leftNode.right = rightNode
    node = node.up

  # Clean up empty levels
  var levelsToRemove: seq[int]
  for i in countdown(len(pt.levels) - 1, 2):
    let levelBelow = pt.levels[i - 1]
    if levelBelow.tail.left.isNil:  # Only has tail node
      levelsToRemove.add(i)
  
  for levelIndex in levelsToRemove:
    pt.levels.delete(levelIndex)

  pt.updatePropagateMerkelHash(rightBoundaryNode.up)
  
  return originalNode

proc `$`*(pt: ProllyTree): string =
  # This procedure will return a string representation of the ProllyTree,
  # showing nodes' timestamps at each level.
  for lvl, level in pt.levels.pairs:
    result.add("Level " & $lvl & ": ")
    var node = level.tail
    # Iterate through nodes at this level from right to left.
    while node != nil:
      result.add($node.timestamp & " ")
      node = node.left  # Move to the next node on the left.
    result.add("\n")  # Add a newline after each level.

proc main() =
  echo "Creating 9 messages: "
  var messages: seq[Message]
  for i in 0..<10:
    messages.add(newMessage($i, i))

  echo "Creating a prolly tree with the messages"
  let tree = newProllyTree(messages)
  echo "Printing the tree"
  echo tree

  echo "#############################################"
  echo "Searching a node with timestamp 5"
  let foundNode = tree.search(5)
  if foundNode != nil:
    echo "Key Found inside the tree: Node(", foundNode.data, ",", foundNode.timestamp, ")"
  else:
    echo "Key not found in the tree."

  echo "#############################################"
  echo "Deleting a node with timestamp 6"
  discard tree.delete(6)
  echo "Printing the tree"
  echo tree

  echo "#############################################"
  echo "Inserting a node with timestamp 12"
  tree.insert(newMessage("12", 12))
  echo "Printing the tree"
  echo tree

when isMainModule:
  main()
